fix encryption header error

jknockaert opened the pull request:
When moving back the pointer to position 0 (using stream_seek), the pointer on the encrypted stream will be moved to the position immediately after the header. Reading the header again (invoked by stream_read) will cause an error, writing the header again (invoked by stream_write) will corrupt the file. Reading/writing the header should therefore happen when opening the file rather than upon read or write. Note that a side-effect of this PR is that empty files will still get an encryption header; I think that is OK, but it is different from how it was originally implemented.
@schiesbn @PVince81 Could you pls check?

Commit:
fix encryption header error

When moving back the pointer to position 0 (using stream_seek), the pointer on the encrypted stream will be moved to the position immediately after the header. Reading the header again (invoked by stream_read) will cause an error, writing the header again (invoked by stream_write) will corrupt the file. Reading/writing the header should therefore happen when opening the file rather than upon read or write. Note that a side-effect of this PR is that empty files will still get an encryption header; I think that is OK, but it is different from how it was originally implemented.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2fe51362-7c1a-4363-bb59-ad8b2793ece0) was created.

jknockaert commented:
While at it I also fixed writing in r+ mode (not sure if that's used anywhere in OC).

Commit:
Update encryption.php

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/78d124de-d030-4118-a92d-62448ee94de5) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/2fe51362-7c1a-4363-bb59-ad8b2793ece0) completed: **6 new issues, 2 updated code elements**

DeepDiver1975 commented:
this results in 56 broken unit tests - https://ci.owncloud.org//job/pull-request-analyser-ng-simple/11712/testReport/

Intended?

jknockaert commented:
No, already fixing.

jknockaert commented:
I hope this the failed tests are now fixed in https://github.com/owncloud/core/commit/0685914c913e600e2ef01766d5d89caf9de2066f. Does Jenkins start retesting automatically?

PVince81 commented:
@jknockaert it does, but if you push before it published the previous results it will become sick.

jknockaert commented:
@PVince81 Is Jenkins sick now? I don't think I pushed before it published...

PVince81 commented:
No, he's still running. Sick is when errors like "could not checkout commit null" or something appears in the log.

PVince81 commented:
Can you add a unit test that reproduces the error and proves that it is fixed ?
Thanks.

jknockaert commented:
@PVince81 OK I'll try to write three tests (r+ mode, read header error, write header error). I'll definitely wait till Jenkins is done with the current one.

jknockaert commented:
@PVince81 Should Jenkins still be dealing with this one? After the last commit to the branch there was the abort message, but Jenkins has been quite since...

schiesbn commented:
@jknockaert after you next commit Jenkins should re-test this PR

Commit:
add two tests

testRewind tests reading and writing after rewind on an encrypted stream; testWriteWriteRead tests r+ mode

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/3ac82aa2-75c2-4c4a-a3d5-a4d51002bcf8) was created.

jknockaert commented:
Not sure if I understand Jenkins, but to me this looks somehow better if not yet perfect.

PVince81 commented:
@owncloud-bot retest this please

Commit:
fixed name

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/28c62c4a-a65c-44e3-8485-aa3347ac86e3) was created.

PVince81 commented:
Okay, so there is indeed a problem with the unit tests:
```
PHP Fatal error:  Call to a member function getHeaderSize() on null in /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/files/stream/encryption.php on line 251
11:52:16 PHP Stack trace:
11:52:16 PHP   1. {main}() /usr/local/bin/phpunit:0
11:52:16 PHP   2. PHPUnit_TextUI_Command::main() /usr/local/bin/phpunit:612
11:52:16 PHP   3. PHPUnit_TextUI_Command->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:138
11:52:16 PHP   4. PHPUnit_TextUI_TestRunner->doRun() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:186
11:52:16 PHP   5. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/TestRunner.php:423
11:52:16 PHP   6. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:751
11:52:16 PHP   7. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:751
11:52:16 PHP   8. PHPUnit_Framework_TestCase->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:751
11:52:16 PHP   9. PHPUnit_Framework_TestResult->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:722
11:52:16 PHP  10. PHP_Invoker->invoke() phar:///usr/local/bin/phpunit/phpunit/Framework/TestResult.php:641
11:52:16 PHP  11. call_user_func_array:{phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93}() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
11:52:16 PHP  12. PHPUnit_Framework_TestCase->runBare() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
11:52:16 PHP  13. PHPUnit_Framework_TestCase->runTest() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:766
11:52:16 PHP  14. ReflectionMethod->invokeArgs() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:881
11:52:16 PHP  15. Test\Files\Stream\Encryption->testStreamOpen() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:881
11:52:16 PHP  16. OC\Files\Stream\Encryption->stream_open() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/files/stream/encryption.php:109
```

Did you try running `./autotest.sh sqlite` locally ?

jknockaert commented:
No. I expected this to be a small&easy fix and implemented it using the github web interface.

Commit:
disable r+ test

jknockaert commented:
I'll test them one by one

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b066bda8-b1d2-4cd5-970f-c62b74d25409) was created.

Commit:
Update encryption.php

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/47bf89c3-1023-406f-a886-a526e5a37ecd) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/3ac82aa2-75c2-4c4a-a3d5-a4d51002bcf8) completed: **4 new issues, 5 updated code elements**

jknockaert commented:
@PVince81 In the last commit all added unit tests were commented out; still Jenkins doesn't like it.

PVince81 commented:
Yes. If you look at the stack trace you can see that it calls `Test\Files\Stream\Encryption->testStreamOpen()` which is another existing test. That test fails because it's calling the code you have changed, which somehow breaks it.

jknockaert commented:
@PVince81 OK thanks. I'll have a look at it.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/28c62c4a-a65c-44e3-8485-aa3347ac86e3) completed: **4 new issues, 7 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b066bda8-b1d2-4cd5-970f-c62b74d25409) completed: **1 new issues, 5 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/47bf89c3-1023-406f-a886-a526e5a37ecd) completed: **3 new issues, 4 updated code elements**

jknockaert commented:
Hmm. Looking at https://github.com/owncloud/core/blob/master/tests/lib/files/stream/encryption.php#L109 it checks for opening a stream with an empty path. Not sure if that should work (even when it does with the existing code). @schiesbn What do you think?

Commit:
Update encryption.php

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/957fab22-9d53-46ef-845d-b62fd0600c9f) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/957fab22-9d53-46ef-845d-b62fd0600c9f) completed: **2 new issues, 2 updated code elements**

Commit:
enable testRewind

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ebed5dc0-9d53-4a94-a1a2-8c08f0564d3a) was created.

Commit:
enable testWriteWriteRead

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/27d01f1a-08e3-4458-a0ae-0a9366fc76a9) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ebed5dc0-9d53-4a94-a1a2-8c08f0564d3a) completed: **3 new issues, 3 updated code elements**

jknockaert commented:
OK, in https://github.com/owncloud/core/commit/0685914c913e600e2ef01766d5d89caf9de2066f I added a hack to avoid testStreamOpen crashing. It looks like my code passes all other tests, including the tests I added. So I suggest that testStreamOpen be reworked; @schiesbn can you advise?

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/27d01f1a-08e3-4458-a0ae-0a9366fc76a9) completed: **1 new issues, 5 updated code elements**

schiesbn commented:
@jknockaert I will have a look at it

schiesbn commented on lib/private/files/stream/encryption.php:
```diff
@@ -238,6 +237,21 @@ public function stream_open($path, $mode, $options, &$opened_path) {
 		$accessList = $this->file->getAccessList($sharePath);
 		$this->newHeader = $this->encryptionModule->begin($this->fullPath, $this->uid, $this->header, $accessList);
 
+		if (!($path==='')){
```
What's the reason for this check? If we want a check like this we want to check $this->fullPath probably. $path is more or less a dummy value

jknockaert replied:
This is a hack to make sure testStreamOpen doesn't crash. I'll remove it after testStreamOpen gets fixed.

schiesbn commented:
This changes should fix it:

````
diff --git a/lib/private/files/stream/encryption.php b/lib/private/files/stream/encryption.php
index 5266027..04bd5ac 100644
--- a/lib/private/files/stream/encryption.php
+++ b/lib/private/files/stream/encryption.php
@@ -237,7 +237,6 @@ class Encryption extends Wrapper {
 		$accessList = $this->file->getAccessList($sharePath);
 		$this->newHeader = $this->encryptionModule->begin($this->fullPath, $this->uid, $this->header, $accessList);
 
-		if (!($path==='')){
 		if (
 			$mode === 'w'
 			|| $mode === 'w+'
@@ -249,13 +248,19 @@ class Encryption extends Wrapper {
 			$this->writeHeader();
 			$this->size = $this->util->getHeaderSize();
 		} else {
-			parent::stream_read($this->util->getHeaderSize());
-		}
+			$this->skipHeader();
 		}
 		return true;
 
 	}
 
+	/**
+	 * read first block to skip the header
+	 */
+	protected function skipHeader() {
+		parent::stream_read($this->util->getHeaderSize());
+	}
+
 	public function stream_eof() {
 		return $this->position >= $this->unencryptedSize;
 	}
@@ -432,7 +437,7 @@ class Encryption extends Wrapper {
 	 * @return integer
 	 * @throws EncryptionHeaderKeyExistsException if header key is already in use
 	 */
-	private function writeHeader() {
+	protected function writeHeader() {
 		$header = $this->util->createHeader($this->newHeader, $this->encryptionModule);
 		return parent::stream_write($header);
 	}
diff --git a/tests/lib/files/stream/encryption.php b/tests/lib/files/stream/encryption.php
index 024185b..0b34de8 100644
--- a/tests/lib/files/stream/encryption.php
+++ b/tests/lib/files/stream/encryption.php
@@ -55,6 +55,7 @@ class Encryption extends \Test\TestCase {
 								   $fileExists,
 								   $expectedSharePath,
 								   $expectedSize,
+								   $expectedUnencryptedSize,
 								   $expectedReadOnly) {
 
 		// build mocks
@@ -77,9 +78,15 @@ class Encryption extends \Test\TestCase {
 				return array();
 			}));
 
+		$utilMock = $this->getMockBuilder('\OC\Encryption\Util')
+			->disableOriginalConstructor()->getMock();
+		$utilMock->expects($this->any())
+			->method('getHeaderSize')
+			->willReturn(8192);
+
 		// get a instance of the stream wrapper
 		$streamWrapper = $this->getMockBuilder('\OC\Files\Stream\Encryption')
-			->setMethods(['loadContext'])->disableOriginalConstructor()->getMock();
+			->setMethods(['loadContext', 'writeHeader', 'skipHeader'])->disableOriginalConstructor()->getMock();
 
 		// set internal properties of the stream wrapper
 		$stream = new \ReflectionClass('\OC\Files\Stream\Encryption');
@@ -95,6 +102,10 @@ class Encryption extends \Test\TestCase {
 		$file->setAccessible(true);
 		$file->setValue($streamWrapper, $fileMock);
 		$file->setAccessible(false);
+		$util = $stream->getProperty('util');
+		$util->setAccessible(true);
+		$util->setValue($streamWrapper, $utilMock);
+		$util->setAccessible(false);
 		$fullPathP = $stream->getProperty('fullPath');
 		$fullPathP->setAccessible(true);
 		$fullPathP->setValue($streamWrapper, $fullPath);
@@ -118,7 +129,7 @@ class Encryption extends \Test\TestCase {
 
 		$unencryptedSize = $stream->getProperty('unencryptedSize');
 		$unencryptedSize->setAccessible(true);
-		$this->assertSame($expectedSize,
+		$this->assertSame($expectedUnencryptedSize,
 			$unencryptedSize->getValue($streamWrapper)
 		);
 		$unencryptedSize->setAccessible(false);
@@ -133,9 +144,9 @@ class Encryption extends \Test\TestCase {
 
 	public function dataProviderStreamOpen() {
 		return array(
-			array('r', '/foo/bar/test.txt', true, '/foo/bar/test.txt', null, true),
-			array('r', '/foo/bar/test.txt', false, '/foo/bar', null, true),
-			array('w', '/foo/bar/test.txt', true, '/foo/bar/test.txt', 0, false),
+			array('r', '/foo/bar/test.txt', true, '/foo/bar/test.txt', null, null, true),
+			array('r', '/foo/bar/test.txt', false, '/foo/bar', null, null, true),
+			array('w', '/foo/bar/test.txt', true, '/foo/bar/test.txt', 8192, 0, false),
 		);
 	}
````

schiesbn commented:
@jknockaert sorry, I thought that you are on your local branch... Give me a second and I will push my fix directly to the branch.

jknockaert commented:
OK. I could also manually apply the diff if you want.

Commit:
fix unit tests

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b3070d83-2a2c-410d-a83a-be190aa8a696) was created.

schiesbn commented:
I applied my fixes and rebased it to current master... Let's see what Jenkins says

jknockaert commented:
OK, looks good. Which one is to go first: this one or #15860?

PVince81 commented:
Looks good :+1: from me and :+1 from @schiesbn who just told me. He'll rebase his.

jknockaert commented:
OK; the branch can deleted I guess.
